README.md

```diff
@@ -120,6 +118,75 @@ describe('MockComponent', () => {
 });
 ```
 
+## MockBuilder
+
+MockBuilder simplifies configuration of TestBed.
+
+```typescript
+import { TestBedStatic } from '@angular/core/testing';
+import { MockBuilder, MockRender } from 'ng-mocks';
+
+describe('MockBuilder', () => {
+  beforeEach(async (done) => {
+    await MockBuilder(MyModule, MyComponent)
+
+      .keep(ModuleWeDontWantToMock)
```

Comment:
Hi, I think that was the point of the conclusion in the issue #44. We need a trigger to know when to reset the cache and without wrapping TestBed I wasn't able to find a way to do it.
With the current implementation every time `.then` is called the cache has been reset.

Reply:
> We need a trigger to know when to reset the cache and without wrapping TestBed I wasn't able to find a way to do it.

Thanks for getting back quickly... If I understand correctly (big if), I think the issue is two-fold:

1. The MockBuilder is operating statically (ie: not using an instance of a builder) so every test shares the same builder which means cleanups across tests must be synchronized. This can be helped by using throw-away instances between tests so cleanup is not necessary.

2. The MockBuilder is directly and automatically firing the TestBed setup which takes some control out of the user's hands when it comes to flow-control. I would recommend making the TestBed configuration step more explicit and allow/require the user trigger it when they are ready. Also, I don't see the need for the `done` argument to the `beforeEach`.

```typescript
beforeEach(async () => {
  await new MockBuilder(FooModule, FooComponent).compileComponents();
});
```

If this all pans out to be compatible with my needs in shallow-render, I'll try to use the MockBuilder myself.

I do see some potential issues with flow-control and mocking in the MockRender design, but my suggestion there would be to just use shallow-render for rendering 😎

Reply:
Hi again.

1. The goal of the #44 is to separate imports and exports with an option of mocking what we need. And the problem is when we want to mock one thing in one test and keep it as is in another with the current implementation we can't do that because for every test we need an own Mock of the same module, that means we need to reset the cache with every TestBed setup. Possible solutions are: to ask user to reset the cache before every testbed or to create an environment when we can detect it ourselves. What I found after investigation is if we wrap the entere setup of TestBed we know when we can reset the cache. That's why MockBuilder has .then which does exactly that - resets cache, mocks everything as we need, sets TestBed and restores original cache.
If we need to add logic when the `TestBed` has been set but the `compileComponents` wasn't called I added one more chaining option: https://take.ms/8fEY3, but it looks a bit dirty to me.

2. Totally agree, and that's what I don't like in the current implementation, but couldn't stand to keep everything in a single chain of setup. A possible option would be to do exactly what we discussed in #44 - just to create a config for TestModule and then to pass it explicitly to the TestBed: https://take.ms/xbVjmg. I'm totally fine with this solution, let's decide and I can update the code.

Reply:
> Possible solutions are: to ask user to reset the cache before every testbed or to create an environment when we can detect it ourselves. 

Just had a thought. You might have success by automatically hooking up a "global" beforeEach that does the cache reset automatically in between tests. That's actually how TestBed works already.

The flow would be:
* TestBed's resetTestingModule fires it it's own beforeEach
* Ng-mock's cache reset fires in it's own beforeEach
* User's beforeEach fires setting up a new MockModule state.

It's a little trick to do it while supporting jest and karma but not too bad.

Here's how I use the same approach to hook up matchers in shallow-render:
https://github.com/getsaf/shallow-render/blob/master/lib/test-frameworks/shallow-matchers.ts#L82

The general idea is that you hook it up automatically when the MockModule js-module is imported, since ja-modules are only interpreted once per-runtime, it only ever hooks up one beforeEach for the whole test suite.

Reply:
a nice hint, thanks! moving to the PR.